client.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2018 Twitch Interactive, Inc.  All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may not
+// use this file except in compliance with the License. A copy of the License is
+// located at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// or in the "license" file accompanying this file. This file is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package twirp
+
+import (
+	"context"
+	"net/http"
+)
+
+type ClientOption func(*ClientOptions)
+
+type ClientOptions struct {
+	Hooks *ClientHooks
+}
+
+type ClientHooks struct {
+	// RequestPrepared is called as soon as a request has been created and before it has been sent
+	// to the Twirp server.
+	RequestPrepared func(context.Context, *http.Request) (context.Context, error)
+
+	// RequestFinished is called after a request has finished sending. Since this is terminal, the context is
+	// not returned.
+	RequestFinished func(context.Context)
```

Comment:
Yeah, I agree, let's rename it. Second question: do we want to pass `res *http.Response` into this function? 🤔Maybe overkill and could be a footgun for people reading the body and stuff in hooks.

Reply:
Fixed in https://github.com/twitchtv/twirp/pull/198/commits/33e3c737ea4b131839e5753a306f73babc28733e.

Reply:
good question, I'm not sure. I could imagine it might be convenient to poke at response headers or status codes, but otoh HTTP-level instrumentation should probably go in a RoundTripper on the HTTPClient.